pkg/mapstreamer/service.go

```diff
@@ -64,43 +64,59 @@ func (fs *Service) MapFn(stream mappb.Map_MapFnServer) error {
 		return err
 	}
 
-	ctx := stream.Context()
+	ctx, cancel := context.WithCancel(stream.Context())
+	defer cancel()
+
+	// Use error group to manage goroutines, the groupCtx is cancelled when any of the
+	// goroutines return an error for the first time or the first time the wait returns.
+	g, groupCtx := errgroup.WithContext(ctx)
+	var readErr error
+	// Read requests from the stream and process them
+outer:
 	for {
-		req, err := recvWithContext(ctx, stream)
+		req, err := recvWithContext(groupCtx, stream)
 		if errors.Is(err, context.Canceled) {
-			log.Printf("Context cancelled, stopping the MapStreamFn")
-			break
+			log.Printf("Context cancelled, stopping the MapFn")
+			break outer
 		}
 		if errors.Is(err, io.EOF) {
-			log.Printf("EOF received, stopping the MapStreamFn")
-			break
+			log.Printf("EOF received, stopping the MapFn")
+			break outer
 		}
-
 		if err != nil {
 			log.Printf("Failed to receive request: %v", err)
-			return err
+			readErr = err
+			// read loop is not part of the error group, so we need to cancel the context
+			// to signal the other goroutines to stop processing.
+			cancel()
+			break outer
 		}
+		g.Go(func() error {
```

Comment:
this is bounded because we control the concurrency on the numaflow side, right?

Reply:
yes